keps/sig-storage/177-volume-snapshot/tighten-validation-webhook-crd.md

```diff
@@ -0,0 +1,815 @@
+
+CRD validation is preferred over webhook validation due to their lower complexity, however CRD validation schema is unable to enforce immutability or provide ratcheting validation.
+
+### Goals
+
+- Provide an updated CRD schema to validate fields
+- Prevent:
+  - Invalid VolumeSnapshot/VolumeSnapshotContent from creation and update
+  - Invalid updates on immutable fields, i.e., VolumeSnapshot.Spec.Source
+- Provide a pre-built image which can be used to deploy the webhook server
+- Provide a way to deploy the webhook server in cluster
+- Provide a way to authenticate the webhook server to the API server via TLS
+- Provide a release process to safely tighten the validation and move towards the ideal state of using builtin CRD validation while maintaining backwards compatibility
+
+### Non-Goals
+
+- Provide a way to authenticate the API server to the webhook server
+
+## Proposal
+
+Tighten the validation on Volume Snapshot objects. The following fields will begin to enforce immutability: `VolumeSnapshot.Spec.Source`, and `VolumeSnapshotContent.Spec.Source`. The following fields will begin to enforce `oneOf`: `VolumeSnapshot.Spec.Source` and `VolumeSnapshotContent.Spec.Source`). The following fields will begin to enforce non-empty strings: `VolumeSnapshot.Spec.VolumeSnapshotClassName`. `VolumeSnapshotContent.spec.VolumeSnapshotRef`will be Immutable only after the UID has been set. More details are in the Validating Scenarios section.
+
+Due to backwards compatibility concerns, the tightening will occur in two phases.
+
+1. The first phase is webhook-only, and will use [ratcheting validation](#backwards-compatibility). It will be the user's responsibility to clean up invalid objects. The controller will not be able to automatically fix invalid objects. The controller will not automatically delete invalid objects to avoid data loss.
+2. The second phase occurs once all invalid objects are cleared from the cluster. The CRD schema validation will be tightened and the webhook will stick around to enforce immutability until immutable fields come to CRDs. (or the crd upgrade could wait until immutable fields are available to do in one go) This will be accompanied by a version change to make it clear the CRD is using different validation.
+
+The phases come in separate releases to allow users / cluster admin the opportunity to clean their cluster of any invalid objects. More details are in the Risks and Mitigations section.
+
+The server will perform validation on Volume Snapshot objects when CREATE and UPDATE requests are made to the api server for `VolumeSnapshot` and `VolumeSnapshotContent` objects. The webhooks will only use validating webhooks, which are read-only.  An image will be built and example `Deployment` and `Service` yaml files will be provided. Example configuration files for the `ValidatingWebhookConfiguration` will be provided, to be used to register the webhooks on the API server.
```

Comment:
"The server ..." - Which one are we referring to?

Reply:
Webhook server.